URL processing options, useful when saving scene to file

janadamec1 opened the pull request:
This adds ProcessUrls procedure with several options in TUrlProcessing:

- suChangeCastleDataToRelative: changes castle-data: URLs to relative paths to target scene filename
- suChangeAllPathsToRelative: changes all paths to relative paths
- suEmbedResources: stores all resources directly inside the scene
- suCopyResourcesToSubdirectory: copies resources to "<name>-resources" subdirectory and stores the reference as relative URL

Usage is implemented in CastleEditor, deprecated_library, and url-processing branch in CastleModelViewer/Converter. The check in GUI, the options are exposed as Radio menu items under Save As... command.

michaliskambi committed:
Define ProcessUrls

- Useful for TDesignFrame.ExportToModel but also other saving routines (like from castle-model-viewer/converter)

- Exposes various ways to process URLs (only 2 are implemented so far: none, and convert casttle-data to relative)

janadamec1 committed:
Fix compilation after UrlProcessing move from CastleViewport to separate source file

janadamec1 committed:
UrlProcessing supports converting to relative filenames and copying to output folder

janadamec1 committed:
Add more X3D nodes for URL processing

janadamec1 committed:
Add URL processing options to deprecated library

janadamec1 committed:
Let CastleEditor compile with UrlProcessing set to previously used value.

janadamec1 committed:
Fix x3dnodes_processurls.inc path inside castle_base package

janadamec1 committed:
Add another URL processing option to convert all paths to relative. Also prepare the subdirectory at processing start for copy-to-subdir option.

janadamec1 committed:
Node.DeepCopy copies its Prototypes too. Needed when using ExternProtos and saving with URL processing.

janadamec1 committed:
Create resources subdirectory only when needed by some url in UrlProcessing

janadamec1 committed:
Embed-resources option for UrlProcessing.

janadamec1 committed:
Keep the bookmark part when processing URLs (may be important for Externprotos)

janadamec1 committed:
Merge branch 'master' into url-processing

michaliskambi commented:
Thank you!

Please add some automatic tests (1. that ProcessUrls works right, 2. that DeepCopy of prototypes works right). Because this is a difficult thing to test, there are various options. Let's use automated testing to help us (and ensure it will keep working forever :) ).

Short primer how to add tests:

You can test running the test suite in "tests/" , see tests/README.md . Short version: compile the project there, and run it :) At the beginning, just run all tests, to make sure it is OK (it should be -- GitHub Actions also check it).

I usually use the "--console" and "--filter xxx" options when developing a new test. So I would add a method like "TestProcessUrls" to TTestX3DNodes (or maybe several methods) and then continue checking

```
cd tests/
castle-engine compile --mode=debug
./castle-tester --console --filter=TTestX3DNodes.TestProcessUrls
```

Of course you can compile and run the testsuite as you like, and use GUI or console runner, whatever is comfortable to you.

The tests don't need to be anything complex, it can be as simple as providing some hardcoded input and looking does it match hardcoded output. Like

```delphi
procedure TTestX3DNodes.TestProcessUrlsEmbed;
begin
  RootNode := TX3DRootNode.Create;
  ...
  ImageTexture.SetUrl('textures/my.png');
  // add ImageTexture to some ShapeNode within RootNode

  TempLocation := GetTempFileName + '.x3d.';
  ProcessUrls(RootNode, TempLocation, suEmbedResources);

  AssertTrue(ImageTexture.Urls.Count = 1);
  AssertTrue(IsSuffix('data:', ImageTexture.Urls[0]));

  finally ... FreeAndNil everyhing :) end;
end;
```

janadamec1 committed:
Fix memory leak in Embedding option of UrlProcessing. Add the test case.

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Fix UrlProcessing test compilation under Delphi

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Fix bad castle_base.lpk merge commit

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

michaliskambi commented:
Sorry for a delay with this, I have this on a list of TODOs for next week!

I see there are 2 conflicts with this PR, if you want to help me you can solve them. But you can also ignore them, I can do them too next week, look trivial :) Just conflicts around LPK / LPI that we indeed changed in CGE master in the meantime.

janadamec1 commented:
No problem, I will merge it into master once the library_update PR is merged (working on it now)

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Fix adding new inc file in base package after merging

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Fix tag in lpk broken in previous merge conflict

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

michaliskambi commented:
I see you're updating this PR, thank you, and thanks a lot for the patience -- I know I have unreasonable delay now with reviewing.

I had busy February (2026), and March (2026) looks busy too, due to some non-computer stuff happening in my life, and also finalizing one job. I have added to TODO, again, to "review this PR" ASAP. I hope to get to it soon. I _should_ have lots more time for CGE work since April 2026, so please bear with me until then :)

P.S. If GHA fails, note that we have now known issue, introduced by me, which makes GHA fail on "master" too. The [PasDoc](https://pasdoc.github.io/) version in [our Docker image](https://castle-engine.io/docker) is now too old. It should be good again in a few hours. (And in positive news, I hope to make PasDoc 1.0.0 release soon :) ).

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Forced commit of fixed line endings

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

janadamec1 committed:
Merge branch 'master' into url-processing

michaliskambi committed:
Merge branch 'master' into url-processing

michaliskambi committed:
Remove line, to remove accidental change - make it closer to master

michaliskambi committed:
Fix apparent mistake when synchronizing url-processing with master (someone solved conflict by copying whole file :) ), bring back ECgeGameControllerButton and ECgeKey mods from master

michaliskambi committed:
Fix finally clause, do it only once RootNodeCopy is initialized; also EInternalError when eUrlProcessing is unexpected (otherwise UrlProcessing has undefined value)

michaliskambi committed:
Typo fix Ember -> Embed

michaliskambi committed:
Avoid warning due to case analysis  ( https://castle-engine.io/coding_conventions#case_analysis )

michaliskambi committed:
Simplify prototypes DeepCopy, passing around NewParentNode was unnecessary - TX3DInterfaceDeclaration inside proto have ParentNode = nil, that's ok

Simplify code and add comments to make it more explicit.

michaliskambi committed:
Add to gitattributes .h and .cpp

michaliskambi committed:
Fix end-of-line normalization for .h, .cpp (following .gitattributes change)

michaliskambi committed:
Autotest DeepCopy of prototypes

michaliskambi committed:
Typo ecgeurlProcessingCopyToOuputDir fix

michaliskambi committed:
Update comment to talk about ProcessUrls

michaliskambi committed:
Extend tests a bit, adding suChangeAllPathsToRelative test

michaliskambi committed:
Small fixes:

- "<name>_resources" document,

- Download never returns nil,

- make try..finally correct (only free F once it is initialized),

- ResourceFileTitle -> ResourceFileBaseName (we never call this "title" in CGE),

- use InclPathDelim (just shorter name).

michaliskambi reviewed (approved):
Thank you again and sorry (for the last time in this PR! :) ) that this took me ages to review.

I've done some improvements, notably: 
- extended tests, to auto-test `DeepCopy` of prototypes
- simplified `DeepCopy` of prototypes -- they don't need `NewParentNode` passed around
- fixed 2 try..finally clauses, to only do `finally` if the freed pointer was initialized

... and I'm happy and merging this now:)

michaliskambi commented:
P.S. I'm aware that logical next step is to refresh branch [url-processing](https://github.com/castle-engine/castle-model-viewer/compare/master...url-processing) of our model viewer (merging master into it, so that it's rebased on master) and applying it too. Added to my TODO. I you will make a PR with it, I will be grateful :)

janadamec1 commented:
Thanks for merging and sorry for that wrong merge of library header.
